services/subscription.go

```diff
@@ -0,0 +1,97 @@
+package services
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/iden3/notification-service/log"
+	"github.com/pkg/errors"
+)
+
+var (
+	// ErrMaxSubscriptionsReached is returned when a user has reached the maximum number of subscriptions
+	ErrMaxSubscriptionsReached = errors.New("maximum number of subscriptions reached")
+)
+
+type Subscriber struct {
+	userDID string
+}
+
+func NewSubscriber(userDID string) Subscriber {
+	return Subscriber{
+		userDID: userDID,
+	}
+}
+
+type SubscriptionService struct {
+	lock        sync.RWMutex
+	subscribers map[Subscriber][]chan NotificationPayload
+
+	maxSubscriptionsPerUser int
+}
+
+func NewSubscriptionService(maxSubscriptionsPerUser int) *SubscriptionService {
+	return &SubscriptionService{
+		lock:        sync.RWMutex{},
+		subscribers: make(map[Subscriber][]chan NotificationPayload),
+
+		maxSubscriptionsPerUser: maxSubscriptionsPerUser,
+	}
+}
+
+func (s *SubscriptionService) Subscribe(userDID string) (<-chan NotificationPayload, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	subscriber := NewSubscriber(userDID)
+	// Check if max subscriptions reached
+	// if maxSubscriptionsPerUser is 0, then unlimited subscriptions are allowed
+	if s.maxSubscriptionsPerUser > 0 &&
+		len(s.subscribers[subscriber]) >= s.maxSubscriptionsPerUser {
+		return nil, fmt.Errorf("%w: allowed connections: %v",
+			ErrMaxSubscriptionsReached, s.maxSubscriptionsPerUser)
+	}
+
+	ch := make(chan NotificationPayload)
```

Comment:
I think it would be nice to make this channel buffered. By design, I think you are going to loop over the list of channels and send message in each channel in the list. In the current unbuffered case, you would lock on each write until somebody is ready to read on the other side (the other size goroutine should be active). If you make it a buffered channel with a buffer size even 1, you just write to this buffer and continue to work unless the buffer is full. Only in this case you would lock instead of each time. So your write goroutine won't wake up and activate some other goroutine on each write and cause a lot of context switches.